server/lib/translator.py

```diff
@@ -21,42 +22,168 @@
+_API_HEADER = {"content-type": "application/json"}
 _EN_LANG = "en"
 
 
 # Detects the query language and translates non-English queries to English.
-def detect_lang_and_translate(query, counters: Counters):
+# Returns a tuple of detected query language and the translated query.
+def detect_lang_and_translate(query, counters: Counters) -> (str, str):
   try:
     lang = lang_detect(query)
     if lang.startswith(_EN_LANG):
-      return query
-
-    api_key = current_app.config.get('PALM_API_KEY')
-    if not api_key:
-      counters.err("translation_api_key_error", "No API key configured.")
-      return query
-
-    request = {"q": query, "format": "text", "source": lang, "target": _EN_LANG}
-    response = requests.post(f'{_API_URL}?key={api_key}',
-                             json=request,
-                             headers=_API_HEADER).json()
-
-    translation = response.get("data",
-                               {}).get("translations",
-                                       [{}])[0].get("translatedText", "")
-    if translation:
+      return lang, query
+
+    translations = _translate([query], lang, _EN_LANG)
+
+    if translations:
       counters.info(
-          "query_translation", {
+          "query_translation",
+          {
               "query": query,
-              "translation": translation,
+              "translation": translations[0],
               "source_language": lang,
-              "target_language": _EN_LANG
-          })
-      return translation
+              "target_language": _EN_LANG,
+          },
+      )
+      return lang, translations[0]
 
     counters.err("empty_query_translation_error", {"query": query})
-    return query
+    return lang, query
   except Exception as e:
     counters.err("query_translation_error", {"query": query, "error": str(e)})
-    return query
+    return _EN_LANG, query
+
+
+class TranslationStringTokens:
+  """Class that replaces tokens (like "${date}") in strings 
+with replacements that won't be translated (like "___") to send for translation.
+
+It also supports a method to reinsert tokens back into the translated string.
+"""
+
+  _TOKEN_PATTERN = r"\${[^}]+}"
+  _TOKEN_REPLACEMENT = "___"
+
+  def __init__(self, original_string: str) -> None:
+    self.original_string = original_string
+    self.tokens = re.findall(TranslationStringTokens._TOKEN_PATTERN,
+                             original_string)
+    self.string_for_translation = re.sub(
+        TranslationStringTokens._TOKEN_PATTERN,
+        TranslationStringTokens._TOKEN_REPLACEMENT,
+        original_string,
+    )
+
+  def reinsert_tokens(self, translation):
+    for token in self.tokens:
+      translation = re.sub(TranslationStringTokens._TOKEN_REPLACEMENT,
+                           token,
+                           translation,
+                           count=1)
+    return translation
+
+
+# Translates display strings in the provided page config and
```

Comment:
There's a lot of logic here that could use some unit tests

Reply:
Done.